near-sdk/src/store/lookup_set/mod.rs

```diff
@@ -1,19 +1,39 @@
 mod impls;
 
 use crate::crypto_hash::{CryptoHasher, Sha256};
-use crate::store::LookupMap;
-use crate::IntoStorageKey;
+use crate::{env, IntoStorageKey, StableMap};
 use borsh::{BorshDeserialize, BorshSerialize};
+use once_cell::unsync::OnceCell;
 use std::borrow::Borrow;
 use std::fmt;
+use std::marker::PhantomData;
+
+const ERR_ELEMENT_SERIALIZATION: &str = "Cannot serialize element";
+
+type LookupKey = [u8; 32];
 
 #[derive(BorshSerialize, BorshDeserialize)]
 pub struct LookupSet<T, H = Sha256>
 where
     T: BorshSerialize + Ord,
     H: CryptoHasher<Digest = [u8; 32]>,
 {
-    map: LookupMap<T, (), H>,
+    prefix: Box<[u8]>,
+
+    /// Cache that keeps track the state of elements in the underlying set.
+    #[borsh_skip]
+    cache: StableMap<T, OnceCell<EntryState>>,
+
+    #[borsh_skip]
+    hasher: PhantomData<H>,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub(crate) enum EntryState {
+    Inserted, // The element is cached as freshly inserted
+    Deleted,  // The element is cached as freshly deleted
+    Present,  // The element is definitely present on the trie
+    Absent,   // The element is definitely absent from the trie
```

Comment:
```suggestion
    /// The element is cached as freshly inserted
    Inserted,
    Deleted,  // The element is cached as freshly deleted
    Present,  // The element is definitely present on the trie
    Absent,   // The element is definitely absent from the trie
```

The standard for this is to have `///` comments above the variants so the descriptions show up with rust tooling or rust docs. Less important since private but probably still worth to change